diff -ur ruby-rpm-1.2.0.orig/ext/rpm/package.c ruby-rpm-1.2.0/ext/rpm/package.c
--- ruby-rpm-1.2.0.orig/ext/rpm/package.c	2006-07-13 13:19:07.000000000 +0200
+++ ruby-rpm-1.2.0/ext/rpm/package.c	2006-07-13 13:29:19.000000000 +0200
@@ -8,6 +8,8 @@
 
 #include "private.h"
 
+char *stpcpy( char *dest, const char *source );
+
 VALUE rpm_cPackage;
 VALUE rpm_sChangeLog;
 
@@ -55,6 +57,8 @@
 	VALUE signature;
 
 	p = Qnil;
+  signature = Qnil;
+
 	if (hdr == NULL) {
 		return Qnil;
 	}
@@ -97,12 +101,13 @@
         	headerAddEntry(hdr,RPMTAG_EPOCH,RPM_INT32_TYPE,&e,1);
 	}
 	pkg = package_new_from_header(klass, hdr);
+  return pkg;
 }
 
 static VALUE
 package_s_open(VALUE klass, VALUE filename)
 {
-	VALUE pkg;
+	VALUE pkg = Qnil;
 	FD_t fd;
 	Header hdr, sigs;
 	rpmRC rc;
@@ -193,6 +198,7 @@
 	for (i=0;i<length;i++)
 		copy_tags[i] = NUM2INT(rb_ary_entry(tags, i));
 	headerCopyTags(RPM_HEADER(from),RPM_HEADER(to),copy_tags);
+  return Qnil;
 }
 
 VALUE
@@ -234,6 +240,7 @@
 	headerAddOrAppendEntry(RPM_HEADER(pkg),nametag,RPM_STRING_ARRAY_TYPE,&name,1);
 	headerAddOrAppendEntry(RPM_HEADER(pkg),versiontag,RPM_STRING_ARRAY_TYPE,&evr,1);
 	headerAddOrAppendEntry(RPM_HEADER(pkg),flagstag,RPM_INT32_TYPE,&flag,1);
+  return Qnil;
 }
 
 VALUE
@@ -251,6 +258,7 @@
 	}
 
 	headerAddOrAppendEntry(RPM_HEADER(pkg),NUM2INT(tag),RPM_INT32_TYPE,&v,1);
+  return Qnil;
 }
 
 VALUE
@@ -260,6 +268,7 @@
 		rb_raise(rb_eTypeError, "illegal argument type");
 	}
 	headerAddOrAppendEntry(RPM_HEADER(pkg),NUM2INT(tag),RPM_STRING_ARRAY_TYPE,&RSTRING(val)->ptr,1);
+  return Qnil;
 }
 
 VALUE
@@ -269,6 +278,7 @@
 		rb_raise(rb_eTypeError, "illegal argument type");
 	}
 	headerAddEntry(RPM_HEADER(pkg),NUM2INT(tag),RPM_STRING_TYPE,RSTRING(val)->ptr,1);
+  return Qnil;
 }
 
 VALUE
@@ -278,6 +288,7 @@
 		rb_raise(rb_eTypeError, "illegal argument type");
 	}
 	headerAddEntry(RPM_HEADER(pkg),NUM2INT(tag),RPM_BIN_TYPE,RSTRING(val)->ptr,RSTRING(val)->len);
+  return Qnil;
 }
 
 VALUE
@@ -335,6 +346,8 @@
 	case RPMTAG_DESCRIPTION:
 		i18n_p = 1;
 		break;
+  default:
+    break;
 	}
 
 	switch (type) {
diff -ur ruby-rpm-1.2.0.orig/ext/rpm/private.h ruby-rpm-1.2.0/ext/rpm/private.h
--- ruby-rpm-1.2.0.orig/ext/rpm/private.h	2006-07-13 13:19:07.000000000 +0200
+++ ruby-rpm-1.2.0/ext/rpm/private.h	2006-07-13 13:29:36.000000000 +0200
@@ -18,6 +18,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
+#include <st.h>
 
 #include <rpmcli.h>
 #include <rpmlib.h>
