diff -ur ruby-rpm-1.2.0/ChangeLog ruby-rpm-1.2.0./ChangeLog
--- ruby-rpm-1.2.0/ChangeLog	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ChangeLog	2004-08-04 21:51:06.000000000 +0900
@@ -1,3 +1,26 @@
+2004-06-05  Kazuhiko  <kazuhiko@fdiary.net>
+
+	* ext/rpm/db.c (rpm_transaction_check): initial value of 'evr' =
+	"" instead of NULL.
+	(package_new_from_NEVR): free 'tmp_nevr' after using 'name'.
+
+2004-06-04  Kazuhiko  <kazuhiko@fdiary.net>
+
+	* ext/rpm/spec.c (rpm_spec_build): add rpmts into the first
+	argument of buildSpec if rpm-4.2.
+
+	* ext/rpm/db.c (rpm_transaction_set_script_file): use
+	rpmtsSetScriptFd instead of rpmtransSetScriptFd if rpm-4.2.
+
+	(rpm_transaction_available): use rpmtsAvailablePackage instead of
+	rpmtransAvailablePackage if rpm-4.2.
+	
+
+2004-06-03  Kazuhiko  <kazuhiko@fdiary.net>
+
+	* ext/rpm/dependency.c (rpm_dependency_is_satisfy): use
+	rpmdsCompare instead of rpmRangesOverlap if rpm-4.2.
+
 2004-05-23  Kazuhiko  <kazuhiko@fdiary.net>
 
 	* tests/test_ts.rb: add 'init_rpmrc()'. build with '--rcfile'
diff -ur ruby-rpm-1.2.0/ext/rpm/db.c ruby-rpm-1.2.0./ext/rpm/db.c
--- ruby-rpm-1.2.0/ext/rpm/db.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/db.c	2004-08-04 21:51:06.000000000 +0900
@@ -4,7 +4,7 @@
  * Copyright (C) 2002 Kenta MURATA <muraken2@nifty.com>.
  */
 
-/* $Id: db.c 27 2004-05-23 04:54:24Z zaki $ */
+/* $Id: db.c 45 2004-06-04 15:11:20Z kazuhiko $ */
 
 #include "private.h"
 
@@ -358,8 +358,10 @@
 	}
 	rb_ivar_set(trans, id_sf, file);
 	RPM_SCRIPT_FD(trans) = fdDup(NUM2INT(rb_Integer(file)));
-#if defined(RPMUPDATE)
+#if RPM_VERSION_CODE < RPM_VERSION(4,1,0)
 	rpmtransSetScriptFd(RPM_TRANSACTION(trans), RPM_SCRIPT_FD(trans));
+#else
+	rpmtsSetScriptFd(RPM_TRANSACTION(trans), RPM_SCRIPT_FD(trans));
 #endif
 	return Qnil;
 }
@@ -446,9 +448,12 @@
 	}
 	rb_ary_push(keys, key);
 
-#if defined(RPMUPDATE)
+#if RPM_VERSION_CODE < RPM_VERSION(4,1,0)
 	rpmtransAvailablePackage(RPM_TRANSACTION(trans), RPM_HEADER(pkg),
 					   RSTRING(key)->ptr);
+#else
+	rpmtsAvailablePackage(RPM_TRANSACTION(trans), RPM_HEADER(pkg),
+					   RSTRING(key)->ptr);
 #endif
 
 	return Qnil;
@@ -557,6 +562,7 @@
 	char *evr = NULL;
 	char *end = NULL;
 	char *tmp_nevr = NULL;
+	VALUE package = Qnil;
 	int i=0;
 
 	tmp_nevr = strdup(nevr);
@@ -577,10 +583,10 @@
 		evr = "";
 	}
 
+	package = rpm_package_new_from_N_EVR(rb_str_new2(name),
+					     version_new_from_EVR(evr));
 	free(tmp_nevr);
-
-	return rpm_package_new_from_N_EVR(rb_str_new2(name),
-									  version_new_from_EVR(evr));
+	return package;
 }
 #endif
 
@@ -663,7 +669,7 @@
 
 				char *name = buf+2;
 				char *relation = NULL;
-				char *evr = NULL;
+				char *evr = "";
 				rpmsenseFlags sense_flags = 0;
 
 				end = strchr ( name, ' ');
@@ -710,8 +716,10 @@
 				break;
 			}
 
+#if 0
 			printf ("%d, type=%d, ignoreProblem=%d, str1=%s pkgNEVR=%s, %s\n",
 					i, p->type, p->ignoreProblem, p->str1, p->pkgNEVR, altNEVR);
+#endif
         }
 	}
 	ps = rpmpsFree(ps);
diff -ur ruby-rpm-1.2.0/ext/rpm/dependency.c ruby-rpm-1.2.0./ext/rpm/dependency.c
--- ruby-rpm-1.2.0/ext/rpm/dependency.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/dependency.c	2004-08-04 21:51:06.000000000 +0900
@@ -4,7 +4,7 @@
  * Copyright (C) 2002 Kenta MURATA <muraken2@nifty.com>.
  */
 
-/* $Id: dependency.c 27 2004-05-23 04:54:24Z zaki $ */
+/* $Id: dependency.c 42 2004-06-03 14:26:37Z kazuhiko $ */
 
 #include "private.h"
 
@@ -196,11 +196,14 @@
 		rb_raise(rb_eTypeError, "illegal argument type");
 	}
 
-#if defined(RPMUPDATE)
+#if RPM_VERSION_CODE < RPM_VERSION(4,1,0)
 	if (rpmRangesOverlap(name,ovre,oflag,
 			     name,svre,NUM2INT(rb_ivar_get(dep, id_flags))))
-		return Qtrue;
+#else
+	if (rpmdsCompare(rpmdsSingle(RPMTAG_PROVIDENAME, name, ovre, oflag),
+			 rpmdsSingle(RPMTAG_PROVIDENAME, name, svre, NUM2INT(rb_ivar_get(dep, id_flags)))))
 #endif
+		return Qtrue;
 	return Qfalse;
 }
 
diff -ur ruby-rpm-1.2.0/ext/rpm/extconf.rb ruby-rpm-1.2.0./ext/rpm/extconf.rb
--- ruby-rpm-1.2.0/ext/rpm/extconf.rb	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/extconf.rb	2004-08-04 21:51:06.000000000 +0900
@@ -21,7 +21,9 @@
 
 def check_db
   dir_config('db')
-
+  if have_library("db-4.2","db_version")
+    return true
+  end
   4.downto(2) do |i|
     if have_library("db-#{i}.0", "db_version") or
         have_library("db#{i}", "db_version") then
diff -ur ruby-rpm-1.2.0/ext/rpm/file.c ruby-rpm-1.2.0./ext/rpm/file.c
--- ruby-rpm-1.2.0/ext/rpm/file.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/file.c	2004-08-04 22:00:35.000000000 +0900
@@ -39,7 +39,7 @@
 	rb_ivar_set(file, id_link_to, RSTRING(link_to)->len ? link_to : Qnil);
 	rb_ivar_set(file, id_size, rb_Integer(size));
 	if (rb_obj_is_kind_of(mtime, rb_cTime) == Qfalse) {
-		mtime = rb_time_new(NUM2INT(rb_Integer(mtime)));
+		mtime = rb_time_new(NUM2INT(rb_Integer(mtime)), (time_t)0);
 	}
 	rb_ivar_set(file, id_mtime, mtime);
 	rb_ivar_set(file, id_owner, owner);
@@ -63,7 +63,7 @@
 	argv[1] = rb_str_new2(md5sum);
 	argv[2] = link_to ? rb_str_new2(link_to) : Qnil;
 	argv[3] = UINT2NUM(size);
-	argv[4] = rb_time_new(mtime);
+	argv[4] = rb_time_new(mtime, (time_t)0);
 	argv[5] = owner ? rb_str_new2(owner) : Qnil;
 	argv[6] = group ? rb_str_new2(group) : Qnil;
 	argv[7] = UINT2NUM(rdev);
diff -ur ruby-rpm-1.2.0/ext/rpm/package.c ruby-rpm-1.2.0./ext/rpm/package.c
--- ruby-rpm-1.2.0/ext/rpm/package.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/package.c	2004-08-04 22:00:55.000000000 +0900
@@ -4,7 +4,7 @@
  * Copyright (C) 2002 Kenta MURATA <muraken2@nifty.com>.
  */
 
-/* $Id: package.c 27 2004-05-23 04:54:24Z zaki $ */
+/* $Id: package.c 45 2004-06-04 15:11:20Z kazuhiko $ */
 
 #include "private.h"
 
@@ -78,7 +78,7 @@
 }
 
 static VALUE
-package_s_create(VALUE klass, VALUE name,VALUE version)
+package_s_create(VALUE klass, VALUE name, VALUE version)
 {
 	Header hdr;
 	VALUE pkg;
@@ -634,7 +634,7 @@
 	for (i = 0; i < count; i++) {
 		VALUE chglog = rb_struct_new(
 			rpm_sChangeLog,
-			rb_time_new(times[i]),
+			rb_time_new(times[i], (time_t)0),
 			rb_str_new2(names[i]),
 			rb_str_new2(texts[i]));
 		rb_ary_push(cl, chglog);
diff -ur ruby-rpm-1.2.0/ext/rpm/rpm.c ruby-rpm-1.2.0./ext/rpm/rpm.c
--- ruby-rpm-1.2.0/ext/rpm/rpm.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/rpm.c	2004-08-04 21:51:06.000000000 +0900
@@ -630,8 +630,6 @@
 	}
 	if (!temp) {
 		temp = "/tmp";
-	} else {
-		temp = strchr(temp, '=') + 1;
 	}
 
 	ruby_rpm_temp_format = rb_str_new2(temp);
diff -ur ruby-rpm-1.2.0/ext/rpm/spec.c ruby-rpm-1.2.0./ext/rpm/spec.c
--- ruby-rpm-1.2.0/ext/rpm/spec.c	2004-05-29 19:14:20.000000000 +0900
+++ ruby-rpm-1.2.0./ext/rpm/spec.c	2004-08-04 21:51:06.000000000 +0900
@@ -4,7 +4,7 @@
  * Copyright (C) 2002 Kenta MURATA <muraken2@nifty.com>.
  */
 
-/* $Id: spec.c 27 2004-05-23 04:54:24Z zaki $ */
+/* $Id: spec.c 44 2004-06-03 16:58:10Z kazuhiko $ */
 
 #include "private.h"
 
@@ -282,8 +282,13 @@
 		rb_raise(rb_eArgError, "argument too many(0..1)");
 	}
 
-#if defined(RPMUPDATE)
+#if RPM_VERSION_CODE < RPM_VERSION(4,1,0)
 	rc = buildSpec(RPM_SPEC(spec), flags,test);
+#else
+	rpmts ts = NULL;
+	ts = rpmtsCreate();
+	rc = buildSpec(ts, RPM_SPEC(spec), flags,test);
+	ts_free(ts);
 #endif
 
 	return INT2NUM(rc);
diff -ur ruby-rpm-1.2.0/tests/test_ts.rb ruby-rpm-1.2.0./tests/test_ts.rb
--- ruby-rpm-1.2.0/tests/test_ts.rb	2004-05-29 19:14:21.000000000 +0900
+++ ruby-rpm-1.2.0./tests/test_ts.rb	2004-08-04 21:51:06.000000000 +0900
@@ -71,6 +71,7 @@
   end
 
   def setup
+    require 'fileutils'
     @work_dir = Dir.tmpdir + "/.ruby-rpm-test"
 
     @rpm_dir = "#{@work_dir}/rpm"
