--- ext/rpm/db.c	2006/03/15 13:50:55	1.1
+++ ext/rpm/db.c	2006/03/15 13:52:11
@@ -746,10 +746,17 @@
 	return rb_ivar_get(trans, id_keys);
 }
 
+#if RPM_VERSION_CODE < RPM_VERSION(4,4,5)
 static void*
 transaction_callback(const void* hd, const rpmCallbackType type,
 					 const unsigned long amount, const unsigned long total,
 					 fnpyKey key, rpmCallbackData data)
+#else
+static void*
+transaction_callback(const void* hd, const rpmCallbackType type,
+					 const unsigned long long amount, const unsigned long long total,
+					 fnpyKey key, rpmCallbackData data)
+#endif
 {
 	VALUE trans = (VALUE)data;
 	FD_t fdt;
